### PR TITLE
[SPARK-50081][SQL] Codegen Support for `XPath*`(by Invoke & RuntimeReplaceable)

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to add `Codegen` Support for `xpath*`, include: 
- `xpath_boolean`
- `xpath_short`
- `xpath_int`
- `xpath_long`
- `xpath_float`
- `xpath_double`
- `xpath_string`
- `xpath`

### Why are the changes needed?
- improve codegen coverage.
- simplified code.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA & Existed UT (eg: `XPathFunctionsSuite`, `XPathExpressionSuite`, `CollationSQLExpressionsSuite`#`*XPath*`, `CollationExpressionWalkerSuite`)

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @MaxGekk @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+object XPathEvaluatorFactory {
+  def create(dataType: DataType, path: UTF8String): XPathEvaluator = {
+    dataType match {
+      case BooleanType => XPathBooleanEvaluator(path)
+      case ShortType => XPathShortEvaluator(path)
+      case IntegerType => XPathIntEvaluator(path)
+      case LongType => XPathLongEvaluator(path)
+      case FloatType => XPathFloatEvaluator(path)
+      case DoubleType => XPathDoubleEvaluator(path)
+      case dt if dt.isInstanceOf[StringType] => XPathStringEvaluator(path)
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:32]: Because we need to support `Collation` (eg: StringType(...)), we have to write it this way.

Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:32]: @panbingkun please add expression-level tests for `xpath_*`

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:32]: > Will it work after your changes? Just in case, are there any tests? cc @stefankandic @uros-db

What I described may have caused `misunderstandings`,
Let me rephrase: This is written so that the case `CollationExpressionWalkerSuite#SPARK-48280: Expression Walker for SQL query examples` can be passed:
https://github.com/apache/spark/blob/be18b7cef3b3162d63930837da6c14469b55759c/sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala#L735-L745

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:32]: > @panbingkun please add expression-level tests (using `checkEvaluation`) for `xpath_*` functions see for example: `CollationStringExpressionsSuite`

Currently, it is `not supported`. `My description` caused a `misunderstanding`, and I have corrected it.

**Review by MaxGekk — COMMENTED** [review]
@panbingkun Could you resolve conflicts, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve conflicts, please.

Updated, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala [hunk 2]
```diff
@@ -40,3 +41,98 @@ object XmlExpressionEvalUtils {
...
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:54]: nit: just for consistency w/ other places:
```suggestion
      case _: StringType => XPathStringEvaluator(path)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala [hunk 3]
```diff
@@ -40,3 +41,98 @@ object XmlExpressionEvalUtils {
...
+      case ArrayType(elementType, _) if elementType.isInstanceOf[StringType] =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:55]:
```suggestion
      case ArrayType(_: StringType, _) =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala [hunk 4]
```diff
@@ -40,3 +41,98 @@ object XmlExpressionEvalUtils {
     UTF8String.fromString(dataType.sql)
   }
 }
+
+object XPathEvaluatorFactory {
+  def create(dataType: DataType, path: UTF8String): XPathEvaluator = {
+    dataType match {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/XmlExpressionEvalUtils.scala:47]: Chasing the `dataType` doesn't look nice. How about to distribute instantiations across expressions:
- Define in `XPathExtract`:
```
  protected def evaluator: XPathEvaluator
``` 
- and override it childs:
```scala
  @transient override lazy val evaluator = XPathBooleanEvaluator(pathUTF8String)
```

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.
